test/Distributed/Distributed.Client/Scenarios/ScenarioRunner.cs

```diff
@@ -0,0 +1,122 @@
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Orleans;
+using Orleans.Configuration;
+using Orleans.Hosting;
+
+namespace Distributed.Client.Scenarios
+{
+    public class CommonParameters
+    {
+        public string ServiceId { get; set; }
+        public string ClusterId { get; set; }
+        public int PipelineSize { get; set; }
+        public int Requests { get; set; } 
+        public int Duration { get; set; }
+        public SecretConfiguration.SecretSource SecretSource { get;set; }
+    }
+
+    public class ScenarioRunner<T>
+    {
+        const int REPORT_BLOCK_SIZE = 100000;
+
+        private readonly IScenario<T> _scenario;
+        private readonly ILogger _logger;
+        private SemaphoreSlim _concurrencyGuard;
+        private volatile int _errorCounter;
+        private volatile int _requestCounter;
+
+        public ScenarioRunner(IScenario<T> scenario, ILoggerFactory loggerFactory)
+        {
+            _scenario = scenario;
+            _logger = loggerFactory.CreateLogger(scenario.Name);
+        }
+
+        public async Task Run(CommonParameters commonParameters, T scenarioParameters)
+        {
+            var requests = commonParameters.Requests != 0 ? commonParameters.Requests : int.MaxValue;
+            var duration = commonParameters.Duration != 0 ? TimeSpan.FromSeconds(commonParameters.Duration) : TimeSpan.MaxValue;
+
+            _logger.LogInformation("Connecting to cluster...");
+            var secrets = SecretConfiguration.Load(commonParameters.SecretSource);
+            var hostBuilder = new HostBuilder()
+                .UseOrleansClient(builder => {
+                    builder
+                        .Configure<ClusterOptions>(options => { options.ClusterId = commonParameters.ClusterId; options.ServiceId = commonParameters.ServiceId; })
+                        .UseAzureStorageClustering(options => options.ConfigureTableServiceClient(secrets.ClusteringConnectionString));
+                });
+            using var host = hostBuilder.Build();
+            await host.StartAsync();
+
+            var client = host.Services.GetService<IClusterClient>();
+
+            _logger.LogInformation("Initializing...");
+            await _scenario.Initialize(client, scenarioParameters, _logger);
+            _concurrencyGuard = new SemaphoreSlim(commonParameters.PipelineSize);
+
+            _logger.LogInformation("Starting load");
+            var cts = new CancellationTokenSource(duration);
+            var stopwatch = Stopwatch.StartNew();
+            for (var i = 0; i < requests; i++)
+            {
+                var request = i;
+                await _concurrencyGuard.WaitAsync();
+                _ = IssueRequest(request);
+                if (cts.IsCancellationRequested)
+                {
+                    break;
+                }
+            }
+            while (_concurrencyGuard.CurrentCount != commonParameters.PipelineSize)
```

Comment:
I think this technique will perform relatively poorly compared to what we have with ConcurrentLoadGenerator https://github.com/dotnet/orleans/blob/main/test/Benchmarks/Ping/ConcurrentLoadGenerator.cs